ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/PeerDiscoveryAgent.java

```diff
@@ -53,282 +51,201 @@
- *       keeping them alive and tracking their state. It is not bothered to service requests.
- * </ol>
+ * messages via UDP.
  */
-public class PeerDiscoveryAgent implements DisconnectCallback {
-  private static final Logger LOG = LogManager.getLogger();
+public abstract class PeerDiscoveryAgent implements DisconnectCallback {
+  protected static final Logger LOG = LogManager.getLogger();
 
   // The devp2p specification says only accept packets up to 1280, but some
   // clients ignore that, so we add in a little extra padding.
   private static final int MAX_PACKET_SIZE_BYTES = 1600;
   private static final long PEER_REFRESH_INTERVAL_MS = MILLISECONDS.convert(30, TimeUnit.MINUTES);
-  private final Vertx vertx;
+
+  protected final List<DiscoveryPeer> bootstrapPeers;
+  private final PeerRequirement peerRequirement;
+  private final PeerBlacklist peerBlacklist;
+  private final NodeWhitelistController nodeWhitelistController;
   /* The peer controller, which takes care of the state machine of peers. */
-  private final PeerDiscoveryController controller;
+  protected Optional<PeerDiscoveryController> controller = Optional.empty();
+
   /* The keypair used to sign messages. */
-  private final SECP256K1.KeyPair keyPair;
+  protected final SECP256K1.KeyPair keyPair;
+  private final BytesValue id;
   private final PeerTable peerTable;
-  private final DiscoveryConfiguration config;
+  protected final DiscoveryConfiguration config;
 
   /* This is the {@link tech.pegasys.pantheon.ethereum.p2p.Peer} object holding who we are. */
   private DiscoveryPeer advertisedPeer;
-  /* The vert.x UDP socket. */
-  private DatagramSocket socket;
+  private InetSocketAddress localAddress;
 
   /* Is discovery enabled? */
   private boolean isActive = false;
+  private final Subscribers<Consumer<PeerBondedEvent>> peerBondedObservers = new Subscribers<>();
 
   public PeerDiscoveryAgent(
-      final Vertx vertx,
       final SECP256K1.KeyPair keyPair,
       final DiscoveryConfiguration config,
       final PeerRequirement peerRequirement,
       final PeerBlacklist peerBlacklist,
       final NodeWhitelistController nodeWhitelistController) {
-    checkArgument(vertx != null, "vertx instance cannot be null");
     checkArgument(keyPair != null, "keypair cannot be null");
     checkArgument(config != null, "provided configuration cannot be null");
 
     validateConfiguration(config);
 
-    final List<DiscoveryPeer> bootstrapPeers =
+    this.peerRequirement = peerRequirement;
+    this.peerBlacklist = peerBlacklist;
+    this.nodeWhitelistController = nodeWhitelistController;
+    this.bootstrapPeers =
         config.getBootstrapPeers().stream().map(DiscoveryPeer::new).collect(Collectors.toList());
 
-    this.vertx = vertx;
     this.config = config;
     this.keyPair = keyPair;
     this.peerTable = new PeerTable(keyPair.getPublicKey().getEncodedBytes(), 16);
-    this.controller =
```

Comment:
interesting, did you remove that circular dependency thing?

Reply:
yeah